File: .local/bin/tmux-sessionizer.go
package main

import (
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
)

type ProjectType string

const (
	Nested ProjectType = "nested"
	Single ProjectType = "single"
)

type Project struct {
	Path string
	Type ProjectType
}

var projects = []Project{
	{Path: "/home/tcrha/dotfiles", Type: Single},
	{Path: "/home/tcrha/Projects", Type: Nested},
	{Path: "/home/tcrha/qbe", Type: Nested},
}

type FolderInfo struct {
	Path   string
	IsOpen bool
}

func main() {
	activeSessions := getActiveTmuxSessions()

	folders := make(map[string]FolderInfo)

	for _, project := range projects {
		if project.Type == Single {
			name := filepath.Base(project.Path)
			sessionName := strings.ReplaceAll(name, ".", "_")
			folders[name] = FolderInfo{
				Path:   project.Path,
				IsOpen: activeSessions[sessionName],
			}
		} else {
			entries, err := os.ReadDir(project.Path)
			if err != nil {
				continue
			}
			for _, entry := range entries {
				if entry.IsDir() {
					fullPath := filepath.Join(project.Path, entry.Name())
					sessionName := strings.ReplaceAll(entry.Name(), ".", "_")
					folders[entry.Name()] = FolderInfo{
						Path:   fullPath,
						IsOpen: activeSessions[sessionName],
					}
				}
			}
		}
	}

	if len(folders) == 0 {
		os.Exit(1)
	}

	var openNames []string
	var closedNames []string
	for name, info := range folders {
		if info.IsOpen {
			openNames = append(openNames, name)
		} else {
			closedNames = append(closedNames, name)
		}
	}

	sort.Strings(openNames)
	sort.Strings(closedNames)

	// Add prefixes after sorting
	var names []string
	for _, name := range openNames {
		names = append(names, "🟢 "+name)
	}
	for _, name := range closedNames {
		names = append(names, "   "+name)
	}

	selected, err := runFzf(names)
	if err != nil {
		os.Exit(1)
	}

	if selected == "" {
		os.Exit(0)
	}

	// Strip the prefix (green circle or spaces) from the selection
	selected = strings.TrimPrefix(selected, "🟢 ")
	selected = strings.TrimPrefix(selected, "   ")

	selectedPath := folders[selected].Path
	selectedName := strings.ReplaceAll(filepath.Base(selected), ".", "_")

	inTmux := os.Getenv("TMUX") != ""
	tmuxRunning := isTmuxRunning()

	if !inTmux && !tmuxRunning {
		// Start new tmux session and attach
		cmd := exec.Command("tmux", "new-session", "-s", selectedName, "-c", selectedPath)
		cmd.Stdin = os.Stdin
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		cmd.Run()
		os.Exit(0)
	}

	// Check if session exists
	if !tmuxSessionExists(selectedName) {
		// Create detached session
		cmd := exec.Command("tmux", "new-session", "-ds", selectedName, "-c", selectedPath)
		cmd.Run()
	}

	if !inTmux {
		// Attach to session
		cmd := exec.Command("tmux", "attach", "-t", selectedName)
		cmd.Stdin = os.Stdin
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		cmd.Run()
	} else {
		// Switch client to session
		cmd := exec.Command("tmux", "switch-client", "-t", selectedName)
		cmd.Run()
	}
}

func runFzf(items []string) (string, error) {
	cmd := exec.Command("fzf", "--layout=reverse", "--tmux", "center", "--ansi")
	cmd.Stdin = strings.NewReader(strings.Join(items, "\n"))
	cmd.Stderr = os.Stderr

	output, err := cmd.Output()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			// fzf returns exit code 130 when user cancels with Ctrl-C
			if exitErr.ExitCode() == 130 || exitErr.ExitCode() == 1 {
				return "", nil
			}
		}
		return "", err
	}

	return strings.TrimSpace(string(output)), nil
}

func isTmuxRunning() bool {
	cmd := exec.Command("pgrep", "tmux")
	err := cmd.Run()
	return err == nil
}

func tmuxSessionExists(name string) bool {
	cmd := exec.Command("tmux", "has-session", "-t="+name)
	err := cmd.Run()
	return err == nil
}

func getActiveTmuxSessions() map[string]bool {
	sessions := make(map[string]bool)

	cmd := exec.Command("tmux", "list-sessions", "-F", "#{session_name}")
	output, err := cmd.Output()
	if err != nil {
		// tmux might not be running, return empty map
		return sessions
	}

	for _, name := range strings.Split(strings.TrimSpace(string(output)), "\n") {
		if name != "" {
			sessions[name] = true
		}
	}

	return sessions
}

File: .local/bin/tmux-sessionizer.ml
open Printf
open Unix

type project_type = Nested | Single

type project = {
  path : string;
  project_type : project_type;
}

type folder_info = {
  path : string;
  is_open : bool;
}

let projects = [
  { path = "/home/tcrha/dotfiles"; project_type = Single };
  { path = "/home/tcrha/qbe"; project_type = Nested };
]

let run_command cmd =
  try
    let ic = Unix.open_process_in cmd in
    let output = ref [] in
    (try
       while true do
         output := input_line ic :: !output
       done
     with End_of_file -> ());
    let _ = Unix.close_process_in ic in
    List.rev !output
  with _ -> []

let command_success cmd =
  try
    let status = Unix.system cmd in
    match status with
    | WEXITED 0 -> true
    | _ -> false
  with _ -> false

let get_active_sessions () =
  let lines = run_command "tmux list-sessions -F '#{session_name}' 2>/dev/null" in
  let sessions = Hashtbl.create 16 in
  List.iter (fun line ->
    if String.length line > 0 then
      Hashtbl.add sessions line true
  ) lines;
  sessions

let is_tmux_running () =
  command_success "pgrep tmux >/dev/null 2>&1"

let tmux_session_exists name =
  command_success (sprintf "tmux has-session -t=%s 2>/dev/null" name)

let run_fzf items =
  try
    let (fd_in, fd_out) = Unix.pipe () in
    let pid = Unix.create_process "fzf"
      [| "fzf"; "--layout=reverse"; "--tmux"; "center"; "--ansi" |]
      fd_in Unix.stdout Unix.stderr in
    Unix.close fd_in;
    let oc = Unix.out_channel_of_descr fd_out in
    List.iter (fun item -> fprintf oc "%s\n" item) items;
    close_out oc;
    let (_, status) = Unix.waitpid [] pid in
    match status with
    | WEXITED 0 ->
      let ic = Unix.open_process_in "fzf --layout=reverse --tmux center --ansi" in
      let temp_file = "/tmp/fzf_input_ocaml" in
      let temp_oc = open_out temp_file in
      List.iter (fun item -> fprintf temp_oc "%s\n" item) items;
      close_out temp_oc;
      let cmd = sprintf "cat %s | fzf --layout=reverse --tmux center --ansi" temp_file in
      let lines = run_command cmd in
      Unix.unlink temp_file;
      (match lines with
       | [] -> None
       | line :: _ -> Some (String.trim line))
    | _ -> None
  with _ -> None

let basename path =
  let parts = String.split_on_char '/' path in
  match List.rev parts with
  | [] -> ""
  | last :: _ -> last

let replace_char s old_char new_char =
  String.map (fun c -> if c = old_char then new_char else c) s

let remove_prefix prefix s =
  let prefix_len = String.length prefix in
  let s_len = String.length s in
  if s_len >= prefix_len && String.sub s 0 prefix_len = prefix then
    String.sub s prefix_len (s_len - prefix_len)
  else
    s

let read_directory path =
  try
    let files = Array.to_list (Sys.readdir path) in
    List.filter (fun name ->
      try Sys.is_directory (Filename.concat path name)
      with _ -> false
    ) files
  with _ -> []

let () =
  printf "tmux-sessionizer ocaml\n";

  let active_sessions = get_active_sessions () in
  let folders = Hashtbl.create 32 in

  List.iter (fun project ->
    match project.project_type with
    | Single ->
      let name = basename project.path in
      let session_name = replace_char name '.' '_' in
      let is_open = Hashtbl.mem active_sessions session_name in
      Hashtbl.add folders name { path = project.path; is_open = is_open }
    | Nested ->
      let entries = read_directory project.path in
      List.iter (fun entry ->
        let full_path = Filename.concat project.path entry in
        let session_name = replace_char entry '.' '_' in
        let is_open = Hashtbl.mem active_sessions session_name in
        Hashtbl.add folders entry { path = full_path; is_open = is_open }
      ) entries
  ) projects;

  if Hashtbl.length folders = 0 then begin
    eprintf "No folders found\n";
    exit 1
  end;

  (* Separate and sort open/closed sessions *)
  let open_names = ref [] in
  let closed_names = ref [] in

  Hashtbl.iter (fun name info ->
    if info.is_open then
      open_names := name :: !open_names
    else
      closed_names := name :: !closed_names
  ) folders;

  let open_names = List.sort String.compare !open_names in
  let closed_names = List.sort String.compare !closed_names in

  (* Build fzf input with prefixes *)
  let names =
    (List.map (fun name -> "🟢 " ^ name) open_names) @
    (List.map (fun name -> "   " ^ name) closed_names) in

  match run_fzf names with
  | None -> exit 0
  | Some selected ->
    (* Strip the prefix *)
    let selected = remove_prefix "🟢 " (remove_prefix "   " selected) in

    let folder_info =
      try Hashtbl.find folders selected
      with Not_found ->
        eprintf "Folder not found\n";
        exit 1 in

    let selected_name = replace_char (basename selected) '.' '_' in
    let in_tmux =
      try String.length (Sys.getenv "TMUX") > 0
      with Not_found -> false in
    let tmux_running = is_tmux_running () in

    if not in_tmux && not tmux_running then begin
      (* Start new tmux session and attach with nvim *)
      let cmd = sprintf "tmux new-session -s %s -c '%s' nvim" selected_name folder_info.path in
      ignore (Unix.system cmd);
      exit 0
    end;

    (* Check if session exists *)
    if not (tmux_session_exists selected_name) then begin
      (* Create detached session with nvim *)
      let cmd = sprintf "tmux new-session -ds %s -c '%s' nvim" selected_name folder_info.path in
      ignore (Unix.system cmd)
    end;

    if not in_tmux then begin
      (* Attach to session *)
      let cmd = sprintf "tmux attach -t %s" selected_name in
      ignore (Unix.system cmd)
    end else begin
      (* Switch client to session *)
      let cmd = sprintf "tmux switch-client -t %s" selected_name in
      ignore (Unix.system cmd)
    end

